docs/adr/0005-Graduation of EventBridge.md

```diff
@@ -0,0 +1,59 @@
+We also plan to support spot instances with AWSMachinePool which would require CAPA controllers to watch for Spot Instance Termination Notices(SITN) which could give enough time to drain spot instances gracefully before they shut down.
+Handle scheduled maintenance events where the EC2 instance needs to be restarted because of reasons like hardware/software updates.
+The AWSInstanceState controller should accommodate different types of events and should be flexible enough to be used by CAPA controllers.
+
+## Decision
+TBD
+
+## Consequences
+### Current Design
+EventBridge looks for EC2 state change events for all the AWSMachine based on InstanceID found in the SQS queue, and loads the state to `ec2-instance-state` label in that AWSMachine in `processMessage()` func.
+
+Currently EventBridge handles messages only from source `aws.ec2`.
+
+### Proposed Design
+* Define `EventBridgeEvent` struct in `awsinstancestate_controller.go` to capture EventBridge event details:
+
+```go
+type EventBridgeEvent struct {
+Version    string          `json:"version"`
+ID         string          `json:"id"`
+DetailType string          `json:"detail-type"`
+Source     string          `json:"source"`
+Account    string          `json:"account"`
+Time       string          `json:"time"`
+Region     string          `json:"region"`
+Resources  []string        `json:"resources"`
+Detail     json.RawMessage `json:"detail"`
+}
+```
+
+* There would be a [long polling mechanism](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-short-and-long-polling.html#sqs-long-polling) in the AWSInstanceState controller that reads the messages from SQS queue with the interval of 10 seconds.
+* `EventBridgeEvent` would be populated based on the unmarshalled message received above.
+* While processing the message in `processMessage()`, AWSInstanceState controller processes below listed events based on the Source captured in above struct:
+   * Check EC2 state change events from `aws.ec2` source.
+   * Check ASG lifecycle hook events from `aws.autoscaling` source.
+   * Check spot instance termination notice events from `aws.ec2` source.
+   * Check scheduled events from `aws.health` source.
+* Based on the type of event, AWSInstanceState controller updates the label ec2-instance-state with the correct instance state.
+* Below helper functions in pkg/cloud/services/instancestate would be used by CAPA controllers:
+   * `GetEventFromSQS()`: This fetches the `EventBridgeEvent` details so as to act upon that event in the respective controller based on the use-case. As soon as the label `ec2-instance-state` is patched on the `AWSMachine`(we would need a watcher for this functionality), this func would be called by the CAPA controller to get the event details.
```

Comment:
We need to think about how to propagate this info to AWSMachinePool controller too. Since ASG instances do not have a matching AWSMachine as of now, there should be additional logic there too.

Reply:
Added.